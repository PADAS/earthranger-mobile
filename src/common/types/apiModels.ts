export interface FileRequest {
  uri: string,
  type: string,
  name: string,
}

export enum ApiStatus {
  Unknown = 1,
  Succeeded = 200,
  BadRequest = 400,
  Unauthorized = 401,
  Forbidden = 403,
  NotFound = 404,
  ServerError = 500,
}
