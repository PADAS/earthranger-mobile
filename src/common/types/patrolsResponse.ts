import { Subject } from './usersResponse';

export interface Status {
  code: number;
  message: string;
}

export interface PatrolType {
  default_priority: number;
  display: string;
  icon_id: string;
  id: string;
  is_active: boolean;
  value: string;
  ordernum: number;
}

export interface PatrolTypesResponse {
  data: PatrolType[];
  status: Status;
}

export interface TrackPatrolByResponse {
  data: {
    properties: {
      leader: {
        enum: Subject [],
      }
    }
  }
  status: Status,
}
