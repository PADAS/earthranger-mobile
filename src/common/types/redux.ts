export interface InitialState {
  session: {
    accessToken: string,
    refreshToken: string,
    error: null,
    sessionError: null,
  },
  reports: {
    noteId: number,
    textNote: string,
  }
}

export type ActionType = 'REFRESH_TOKEN'
| 'REFRESH_TOKEN_ERROR'
| 'USER_LOGIN'
| 'USER_LOGIN_ERROR'
| 'GET_SESSION'
| 'GET_SESSION_ERROR'
| 'CREATE_NOTE'
| 'CLEAR_NOTE_DATA'
| 'NUMBER_OF_REPORT_DRAFTS'
| 'CLEAR_SESSION';

export interface ReducerAction {
  type: ActionType,
  payload: any;
}
