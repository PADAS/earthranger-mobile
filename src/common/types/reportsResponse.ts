export interface ReportCategoriesResponse {
  data: Category[];
  status: Status;
}

export interface Category {
  id: string;
  value: string;
  display: string;
  ordernum: number;
  flag?: string;
  account_id: string | undefined;
  profile_id: number[];
}

export interface ReportTypesResponse {
  data: Type[];
  status: Status;
}

export interface ReportTypeSchemaResponse {
  data: string;
  status: Status;
}

export interface Type {
  id: string;
  icon: string;
  value: string;
  display: string;
  ordernum: number;
  is_collection: boolean;
  readonly?: boolean;
  category: Category;
  icon_id: string;
  is_active: boolean;
  default_priority: number;
  geometry_type: string;
  schema: string;
  url: string;
  icon_svg: string;
}

export interface Status {
  code: number;
  message: string;
}

export interface EventCategory {
  id: number;
  remote_id: string;
  value: string;
  display: string;
}

export interface EventType {
  id: number;
  display: string;
  icon_svg: string;
  default_priority: string;
}

export interface Note {
  id: number,
  text: string,
}

export interface WithChildren {
  children: any;
}

export interface UploadedReport {
  id: number;
}
