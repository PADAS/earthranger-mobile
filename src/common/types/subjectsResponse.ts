export interface SubjectGroupResponse {
  data: SubjectGroup[],
  status: Status,
}

export interface SubjectGroup {
  name: string,
  id: string,
  subgroups: SubjectGroup[],
  subjects: Subject[],
}

export interface Subject {
  content_type: string,
  id: string,
  name: string,
  subject_type: string,
  subject_subtype: string,
  common_name: string | null,
  additional: SubjectAdditional,
  created_at: string,
  updated_at: string,
  is_active: boolean,
  user: null,
  region: string,
  country: string,
  sex: string,
  tracks_available: boolean,
  image_url: string,
  last_position_status: SubjectLastPositionStatus,
  last_position_date: string,
  last_position: SubjectLastPosition,
  device_status_properties: SubjectDeviceStatusProperties[],
  url: string,
}

interface SubjectAdditional {
  rgb: string,
  sex: string,
  region: string,
  country: string,
  external_id: string,
  tm_animal_id: string,
  external_name: string,
}

interface SubjectLastPositionStatus {
  last_voice_call_start_at: string | null,
  radio_state_at: string | null,
  radio_state: string,
}

interface SubjectLastPosition {
  type: string,
  geometry: {
    type: string,
    coordinates: [number, number]
  },
  properties: {
    title: string,
    subject_type: string,
    subject_subtype: string,
    id: string,
    stroke: string,
    stroke_opacity: number,
    stroke_width: number,
    image: string,
    last_voice_call_start_at: string | null,
    location_requested_at: string | null,
    radio_state_at: string,
    radio_state: string,
    coordinateProperties: {
      time: string,
    },
    DateTime: string,
  }
}

interface SubjectDeviceStatusProperties {
  label: string,
  units: string,
  value: number,
}

interface Status {
  code: number;
  message: string;
}
