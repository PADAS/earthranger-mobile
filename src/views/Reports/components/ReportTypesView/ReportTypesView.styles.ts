// External Dependencies
import { StyleSheet } from 'react-native';

// Internal Dependencies
import { COLORS_LIGHT } from '../../../../common/constants/colors';
import { IS_IOS } from '../../../../common/constants/constants';

const styles = StyleSheet.create({
  content: {
    flex: 1,
    backgroundColor: COLORS_LIGHT.white,
  },
  flatList: {
    zIndex: 0,
  },
  columnWrapperStyle: {
    flex: 1,
    alignItems: 'stretch',
  },
  searchInput: {
    flex: 1,
    marginStart: IS_IOS ? 10 : 0,
    marginEnd: IS_IOS ? 80 : 40,
    fontSize: 17,
  },
  searchBackIcon: {
    marginRight: IS_IOS ? 0 : 16,
  },
});

export default styles;
