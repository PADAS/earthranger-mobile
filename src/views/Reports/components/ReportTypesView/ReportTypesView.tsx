// External Dependencies
import React, {
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  TextInput,
} from 'react-native';
import { FlashList } from '@shopify/flash-list';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { RouteProp } from '@react-navigation/native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { useTranslation } from 'react-i18next';
import { View } from 'react-native-ui-lib';

// Internal Dependencies
import { COLORS_LIGHT } from '../../../../common/constants/colors';
import { cropHeaderTitleText } from '../../../../common/utils/stringUtils';
import { customBackButton, osBackIcon } from '../../../../common/components/header/header';
import { EmptySearchResultsView } from '../../../../common/components/EmptySearchResults/EmptySearchResultsView';
import { EventType } from '../../../../common/types/reportsResponse';
import { getBoolForKey, setStringForKey } from '../../../../common/data/storage/keyValue';
import {
  ELEMENT_INSPECTOR_WIDTH,
  MERGE_CATEGORIES_KEY,
  PATROL_INFO_EVENT_TYPE_VALUE,
} from '../../../../common/constants/constants';
import { PermissionView } from '../../../Permission/PermissionView';
import { PersistedEventType, Position, RootStackParamList } from '../../../../common/types/types';
import { ReportTypesCell } from './components/ReportTypesCell/ReportTypesCell';
import { SearchButton } from '../../../../common/components/SearchButton/SearchButton';
import { useFilterReportTypesByDisplay } from '../../../../common/data/reports/useFilterReportTypesByDisplay';
import { useRetrieveEventPermissions } from '../../../../common/data/permissions/useRetrievePermissions';
import { useRetrieveReportTypes } from '../../../../common/data/reports/useRetrieveReportTypes';
import { useRetrieveReportTypesByCategory } from '../../../../common/data/reports/useRetrieveReportTypesByCategory';
import { useRetrieveUser } from '../../../../common/data/users/useRetrieveUser';
import { UserType } from '../../../../common/enums/enums';

// Styles
import styles from './ReportTypesView.styles';

// Interfaces + Types
interface EventTypesViewProps {
  navigation: NativeStackNavigationProp<RootStackParamList, 'ReportTypesView'>;
  route: RouteProp<{
    params: {
      categoryId: string,
      coordinates?: Position,
      isPatrolInfoEventType?: boolean,
      title: string,
    }
  },
  'params'>,
}

const ReportTypesView = ({ route, navigation }: EventTypesViewProps) => {
  // Hooks
  const { filterReportTypesByDisplay } = useFilterReportTypesByDisplay();
  const { retrieveReportTypesByCategory } = useRetrieveReportTypesByCategory();
  const { retrieveReportTypesByUserType } = useRetrieveReportTypes();
  const { retrieveUserInfo } = useRetrieveUser();
  const { t } = useTranslation();
  const { userHasEventsPermissions } = useRetrieveEventPermissions();

  // References
  const inputRef = useRef<TextInput>(null);
  const mergeCategories = useRef(
    route.params.isPatrolInfoEventType || getBoolForKey(MERGE_CATEGORIES_KEY),
  );
  const isSearching = useRef(false);

  // State
  const [coordinates, setCoordinates] = useState<Position>(route.params.coordinates || [0, 0]);
  const [emptyEventCategory, setEmptyEventCategory] = useState(false);
  const [eventTypes, setEventTypes] = useState<EventType[] | PersistedEventType[]>([]);
  const [filteredEventTypes, setFilteredEventTypes] = useState<EventType[]>([]);
  const [isSearchModeEnabled, setIsSearchModeEnabled] = useState(false);
  const [showPermissionView, setShowPermissionView] = useState(false);

  // Lifecycle Events
  useEffect(() => {
    initAppBar();
    initCoordinates();
    getEventTypes();
  }, []);

  useEffect(() => {
    if ((isSearching.current && filteredEventTypes.length > 0)
      || (!isSearching.current && eventTypes.length > 0)) {
      setEventTypes(isSearching.current ? filteredEventTypes : eventTypes);
    }
  }, [isSearching, eventTypes, filteredEventTypes]);

  useEffect(() => {
    if (isSearchModeEnabled && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isSearchModeEnabled]);

  useEffect(() => {
    if (showPermissionView) {
      hideSearchIcon();
    }
  }, [showPermissionView]);

  // Utility Functions

  const initAppBar = () => {
    let viewTitle = t('reports.appBarTitle');

    if (route.params.title) {
      viewTitle = cropHeaderTitleText(route.params.title);
    }

    navigation.setOptions({
      // @ts-ignore
      title: viewTitle,
      headerTitle: viewTitle,
      headerLeft: () => customBackButton(
        osBackIcon,
        () => navigation.pop(),
        getBoolForKey(MERGE_CATEGORIES_KEY),
      ),
    });

    // Workaround found in https://github.com/software-mansion/react-native-screens/issues/1570#issuecomment-1362794902
    setTimeout(() => {
      navigation.setOptions({
        // @ts-ignore
        headerRight: () => searchIconView(),
      });
    }, 10);
  };

  const initCoordinates = useCallback(() => {
    if (route.params.coordinates) {
      setCoordinates(
        [
          parseFloat(route.params.coordinates[0].toFixed(6)),
          parseFloat(route.params.coordinates[1].toFixed(6)),
        ],
      );
    }
  }, [route.params.coordinates]);

  const getEventTypes = async () => {
    let eventTypesList: EventType[] | PersistedEventType[] = [];
    const hasEventsPermission = await userHasEventsPermissions();

    if (mergeCategories.current) {
      const userInfo = await retrieveUserInfo();

      if (userInfo?.userType) {
        const profileId = userInfo?.userType === UserType.profile ? parseInt(userInfo.userId || '', 10) : undefined;

        eventTypesList = await retrieveReportTypesByUserType(
          userInfo.userType,
          userInfo?.permissions,
          profileId,
        );

        if (eventTypesList.length === 0) {
          setShowPermissionView(true);
          setEmptyEventCategory(hasEventsPermission);
        }

        setEventTypes(eventTypesList);
      }
    } else {
      eventTypesList = await retrieveReportTypesByCategory([route.params.categoryId]);

      if (eventTypesList.length === 0) {
        setShowPermissionView(true);
        setEmptyEventCategory(hasEventsPermission);
      }

      setEventTypes(eventTypesList);
    }
  };

  const updateAppBar = async () => {
    navigation.setOptions({
      // @ts-ignore
      headerLeft: () => searchBackIconView(),
      headerRight: () => null,
      headerTitle: () => searchInputView(),
      headerBackVisible: false,
    });
  };

  const hideSearchIcon = () => {
    navigation.setOptions({
      // @ts-ignore
      headerRight: () => null,
    });
  };

  const filterEventTypes = async (queryText: string) => {
    const filteredEventTypesList = await filterReportTypesByDisplay(
      [
        route.params.categoryId,
        `%${queryText}%`,
      ],
      mergeCategories.current,
      queryText,
    );
    setFilteredEventTypes(filteredEventTypesList);
  };

  const showSearchInput = async () => {
    await updateAppBar();
    setIsSearchModeEnabled(true);
  };

  const backFromSearch = () => {
    initAppBar();
    getEventTypes();
    isSearching.current = false;
    setIsSearchModeEnabled(false);
  };

  const searchEvent = (text: string) => {
    isSearching.current = true;
    filterEventTypes(text);
  };

  const navigateToEventForm = useCallback((
    titleValue: string,
    idValue: string,
    schema: string,
    geometryType: string,
  ) => {
    navigation.navigate('ReportForm', {
      coordinates,
      geometryType,
      schema,
      title: titleValue,
      typeId: idValue,
    });
  }, [coordinates]);

  const onEventPressHanlder = useCallback((item) => {
    if (route.params.isPatrolInfoEventType) {
      setStringForKey(PATROL_INFO_EVENT_TYPE_VALUE, item.value);
      navigation.goBack();
    } else {
      navigateToEventForm(
        item.display,
        item.id.toString(),
        item.schema,
        item.geometry_type,
      );
    }
  }, []);

  // Additional Components

  const searchIconView = () => (<SearchButton onPress={showSearchInput} />);

  const searchBackIconView = () => customBackButton(osBackIcon, backFromSearch);

  const searchInputView = () => (
    <TextInput
      placeholder={t('reportTypes.searchPlaceholder')}
      style={styles.searchInput}
      onChangeText={(text) => (searchEvent(text))}
      ref={inputRef}
      placeholderTextColor={COLORS_LIGHT.G3_secondaryMediumLightGray}
    />
  );

  // List components

  const renderItem = ({ item }: any) => (
    <ReportTypesCell
      iconImage={item.icon_svg}
      onPress={() => onEventPressHanlder(item)}
      priority={item.default_priority}
      title={item.display}
      typeId={item.id.toString()}
    />
  );

  return (
    <SafeAreaView style={styles.content} edges={['bottom']}>
      {(!showPermissionView) ? (
        <View>
          {(isSearching.current && filteredEventTypes.length === 0)
            ? <EmptySearchResultsView />
            : (
              <View style={{ height: '100%' }}>
                <FlashList
                  // @ts-ignore
                  data={eventTypes}
                  estimatedItemSize={ELEMENT_INSPECTOR_WIDTH}
                  keyboardShouldPersistTaps="always"
                  keyExtractor={(item) => item.id.toString()}
                  numColumns={3}
                  renderItem={renderItem}
                />
              </View>
            )}
        </View>
      ) : (
        <PermissionView emptyEventCategory={emptyEventCategory} />
      )}
    </SafeAreaView>
  );
};

export { ReportTypesView };
