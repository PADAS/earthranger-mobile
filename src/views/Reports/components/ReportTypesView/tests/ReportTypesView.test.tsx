// External Dependencies
import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { RootStackParamList } from '../../../../../common/types/types';

// Internal Dependencies
import { ReportTypesView } from '../ReportTypesView';

// Types + Interfaces
// eslint-disable-next-line max-len
type StackNavigationPropAlias = NativeStackNavigationProp<RootStackParamList, 'ReportTypesView'>;

// Test Specific Mocks
jest.mock('../../../../../common/data/reports/useRetrieveReportTypesByCategory', () => ({
  useRetrieveReportTypesByCategory: () => ({
    retrieveReportTypesByCategory: () => Promise.resolve([
      {
        id: 1,
        display: 'Fence Report',
        priority: '100',
        icon_svg: '',
      },
      {
        id: 2,
        display: 'Crossing Report',
        priority: '200',
        icon_svg: '',
      },
      {
        id: 3,
        display: 'Confiscation',
        priority: '300',
        icon_svg: '',
      },
    ]),
  }),
}));

jest.mock('react-native-geolocation-service', () => ({
  getCurrentPosition: () => ({
    position: {
      coords: {
        latitude: 123.456,
        longitude: 456.123,
      },
    },
  }),
}));

jest.mock('recyclerlistview', () => ({
  ...jest.requireActual('recyclerlistview'),
  RecyclerListView: ({ dataProvider, layoutProvider, rowRenderer }) => {
    const data = dataProvider.getAllData();
    return data.map((datum, index) => rowRenderer(
      layoutProvider.getLayoutTypeForIndex(index),
      datum,
    ));
  },
}));

describe('ReportTypesView', () => {
  let navigation: Pick<StackNavigationPropAlias, 'setOptions' >;
  const route = {
    params: {
      category: '1',
      coordinates: [0, 0],
    },
  };

  beforeEach(() => {
    navigation = {
      setOptions: jest.fn(),
    };
  });

  it('matches snapshot', async () => {
    const { toJSON, getByText } = render(
      // @ts-ignore
      <ReportTypesView navigation={navigation} route={route} />,
    );

    await waitFor(() => {
      expect(getByText('Fence Report')).toBeTruthy();
    });

    expect(toJSON()).toMatchSnapshot();
  });
});
