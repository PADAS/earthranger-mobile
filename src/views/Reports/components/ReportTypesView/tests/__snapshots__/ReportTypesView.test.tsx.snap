// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportTypesView matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Object {
            "height": "100%",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "overflow": "hidden",
          }
        }
      >
        <RCTScrollView
          applyWindowCorrection={[Function]}
          canChangeSize={true}
          contentContainerStyle={
            Object {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
              "paddingLeft": 0,
              "paddingRight": 0,
            }
          }
          contentHeight={140}
          contentWidth={400}
          data={
            Array [
              Object {
                "display": "Fence Report",
                "icon_svg": "",
                "id": 1,
                "priority": "100",
              },
              Object {
                "display": "Crossing Report",
                "icon_svg": "",
                "id": 2,
                "priority": "200",
              },
              Object {
                "display": "Confiscation",
                "icon_svg": "",
                "id": 3,
                "priority": "300",
              },
            ]
          }
          dataProvider={
            DataProvider {
              "_data": Array [
                Object {
                  "display": "Fence Report",
                  "icon_svg": "",
                  "id": 1,
                  "priority": "100",
                },
                Object {
                  "display": "Crossing Report",
                  "icon_svg": "",
                  "id": 2,
                  "priority": "200",
                },
                Object {
                  "display": "Confiscation",
                  "icon_svg": "",
                  "id": 3,
                  "priority": "300",
                },
              ],
              "_firstIndexToProcess": 0,
              "_hasStableIds": true,
              "_requiresDataChangeHandling": false,
              "_size": 3,
              "getStableId": [Function],
              "rowHasChanged": [Function],
            }
          }
          disableRecycling={false}
          estimatedItemSize={140}
          extendedState={Object {}}
          externalScrollView={[Function]}
          finalRenderAheadOffset={250}
          forceNonDeterministicRendering={true}
          horizontal={false}
          initialOffset={0}
          initialRenderIndex={0}
          isHorizontal={false}
          keyExtractor={[Function]}
          keyboardShouldPersistTaps="always"
          layoutProvider={
            GridLayoutProviderWithProps {
              "_acceptableRelayoutDelta": 1,
              "_getHeightOrWidth": [Function],
              "_getLayoutTypeForIndex": [Function],
              "_getSpan": [Function],
              "_hasExpired": false,
              "_isHorizontal": false,
              "_lastLayoutManager": GridLayoutManager {
                "_acceptableRelayoutDelta": 1,
                "_getSpan": [Function],
                "_isGridHorizontal": false,
                "_isHorizontal": false,
                "_layoutProvider": [Circular],
                "_layouts": Array [
                  Object {
                    "height": 140,
                    "type": 0,
                    "width": 133.33333333333334,
                    "x": 0,
                    "y": 0,
                  },
                  Object {
                    "height": 140,
                    "type": 0,
                    "width": 133.33333333333334,
                    "x": 133.33333333333334,
                    "y": 0,
                  },
                  Object {
                    "height": 140,
                    "type": 0,
                    "width": 133.33333333333334,
                    "x": 266.6666666666667,
                    "y": 0,
                  },
                ],
                "_maxSpan": 3,
                "_renderWindowSize": Object {
                  "height": 900,
                  "width": 400,
                },
                "_totalHeight": 140,
                "_totalWidth": 400,
                "_window": Object {
                  "height": 900,
                  "width": 400,
                },
              },
              "_maxSpan": 3,
              "_renderWindowSize": Object {
                "height": 900,
                "width": 400,
              },
              "_setLayoutForType": [Function],
              "_tempDim": Object {
                "height": 0,
                "width": 0,
              },
              "averageWindow": AverageWindow {
                "currentAverage": 140,
                "currentCount": 1,
                "inputValues": Array [
                  140,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                  ,
                ],
                "nextIndex": 1,
              },
              "defaultEstimatedItemSize": 100,
              "layoutObject": Object {
                "size": undefined,
                "span": undefined,
              },
              "props": Object {
                "data": Array [
                  Object {
                    "display": "Fence Report",
                    "icon_svg": "",
                    "id": 1,
                    "priority": "100",
                  },
                  Object {
                    "display": "Crossing Report",
                    "icon_svg": "",
                    "id": 2,
                    "priority": "200",
                  },
                  Object {
                    "display": "Confiscation",
                    "icon_svg": "",
                    "id": 3,
                    "priority": "300",
                  },
                ],
                "estimatedItemSize": 140,
                "keyExtractor": [Function],
                "keyboardShouldPersistTaps": "always",
                "numColumns": 3,
                "renderItem": [Function],
              },
              "renderWindowInsets": Object {
                "height": 0,
                "width": 0,
              },
              "shouldRefreshWithAnchoring": true,
            }
          }
          maxRenderAhead={750}
          numColumns={3}
          onEndReached={[Function]}
          onEndReachedThreshold={0}
          onEndReachedThresholdRelative={0}
          onItemLayout={[Function]}
          onLayout={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onSizeChanged={[Function]}
          onVisibleIndicesChanged={[Function]}
          removeClippedSubviews={false}
          renderAheadOffset={0}
          renderAheadStep={250}
          renderContentContainer={[Function]}
          renderItem={[Function]}
          renderItemContainer={[Function]}
          rowRenderer={[Function]}
          scrollEventThrottle={16}
          scrollThrottle={16}
          scrollViewProps={
            Object {
              "contentContainerStyle": Object {
                "backgroundColor": undefined,
                "minHeight": 1,
                "minWidth": 1,
                "paddingLeft": 0,
                "paddingRight": 0,
              },
              "onLayout": [Function],
              "onScrollBeginDrag": [Function],
              "refreshControl": undefined,
              "style": Object {
                "minHeight": 1,
                "minWidth": 1,
              },
            }
          }
          style={
            Object {
              "minHeight": 1,
              "minWidth": 1,
            }
          }
          suppressBoundedSizeException={true}
          windowCorrectionConfig={
            Object {
              "applyToInitialOffset": true,
              "applyToItemScroll": true,
              "value": Object {
                "endCorrection": 0,
                "startCorrection": 0,
                "windowShift": -0,
              },
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "paddingLeft": 0,
                    "paddingTop": 0,
                  }
                }
              />
              <View
                style={
                  Array [
                    undefined,
                    undefined,
                  ]
                }
              />
              <AutoLayoutView
                enableInstrumentation={false}
                horizontal={false}
                onBlankAreaEvent={[Function]}
                onLayout={[Function]}
                renderAheadOffset={0}
                scrollOffset={0}
                style={
                  Object {
                    "height": 140,
                    "width": 400,
                  }
                }
                windowSize={900}
              >
                <CellContainer
                  index={0}
                  onLayout={[Function]}
                  style={
                    Object {
                      "alignItems": "stretch",
                      "flexDirection": "column",
                      "left": 0,
                      "position": "absolute",
                      "top": 0,
                      "width": 133.33333333333334,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                      }
                    }
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "opacity": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "borderBottomWidth": 1,
                            "borderColor": "#DDDDDD",
                            "borderRightWidth": 1,
                            "flexDirection": "column",
                            "height": 131,
                            "justifyContent": "center",
                            "paddingBottom": 8,
                            "paddingLeft": 8,
                            "paddingRight": 8,
                            "paddingTop": 16,
                            "width": 250,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "display": "flex",
                              "height": 40,
                              "justifyContent": "center",
                              "marginBottom": 16,
                            }
                          }
                        >
                          <RNSVGSvgView
                            bbHeight="24"
                            bbWidth="24"
                            focusable={false}
                            height="24"
                            style={
                              Array [
                                Object {
                                  "backgroundColor": "transparent",
                                  "borderWidth": 0,
                                },
                                Object {
                                  "flex": 0,
                                  "height": 24,
                                  "width": 24,
                                },
                              ]
                            }
                            width="24"
                          >
                            <RNSVGGroup>
                              <RNSVGGroup>
                                <RNSVGCircle
                                  cx={12}
                                  cy={12}
                                  fill={4278190080}
                                  propList={
                                    Array [
                                      "fill",
                                    ]
                                  }
                                  r={12}
                                />
                              </RNSVGGroup>
                            </RNSVGGroup>
                          </RNSVGSvgView>
                        </View>
                        <Text
                          numberOfLines={3}
                          style={
                            Object {
                              "fontSize": 15,
                              "height": 54,
                              "lineHeight": 17,
                              "textAlign": "center",
                            }
                          }
                        >
                          Fence Report
                        </Text>
                      </View>
                    </View>
                  </View>
                </CellContainer>
                <CellContainer
                  index={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "alignItems": "stretch",
                      "flexDirection": "column",
                      "left": 133.33333333333334,
                      "position": "absolute",
                      "top": 0,
                      "width": 133.33333333333334,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                      }
                    }
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "opacity": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "borderBottomWidth": 1,
                            "borderColor": "#DDDDDD",
                            "borderRightWidth": 1,
                            "flexDirection": "column",
                            "height": 131,
                            "justifyContent": "center",
                            "paddingBottom": 8,
                            "paddingLeft": 8,
                            "paddingRight": 8,
                            "paddingTop": 16,
                            "width": 250,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "display": "flex",
                              "height": 40,
                              "justifyContent": "center",
                              "marginBottom": 16,
                            }
                          }
                        >
                          <RNSVGSvgView
                            bbHeight="24"
                            bbWidth="24"
                            focusable={false}
                            height="24"
                            style={
                              Array [
                                Object {
                                  "backgroundColor": "transparent",
                                  "borderWidth": 0,
                                },
                                Object {
                                  "flex": 0,
                                  "height": 24,
                                  "width": 24,
                                },
                              ]
                            }
                            width="24"
                          >
                            <RNSVGGroup>
                              <RNSVGGroup>
                                <RNSVGCircle
                                  cx={12}
                                  cy={12}
                                  fill={4278190080}
                                  propList={
                                    Array [
                                      "fill",
                                    ]
                                  }
                                  r={12}
                                />
                              </RNSVGGroup>
                            </RNSVGGroup>
                          </RNSVGSvgView>
                        </View>
                        <Text
                          numberOfLines={3}
                          style={
                            Object {
                              "fontSize": 15,
                              "height": 54,
                              "lineHeight": 17,
                              "textAlign": "center",
                            }
                          }
                        >
                          Crossing Report
                        </Text>
                      </View>
                    </View>
                  </View>
                </CellContainer>
                <CellContainer
                  index={2}
                  onLayout={[Function]}
                  style={
                    Object {
                      "alignItems": "stretch",
                      "flexDirection": "column",
                      "left": 266.6666666666667,
                      "position": "absolute",
                      "top": 0,
                      "width": 133.33333333333334,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                      }
                    }
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "opacity": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "borderBottomWidth": 1,
                            "borderColor": "#DDDDDD",
                            "borderRightWidth": 1,
                            "flexDirection": "column",
                            "height": 131,
                            "justifyContent": "center",
                            "paddingBottom": 8,
                            "paddingLeft": 8,
                            "paddingRight": 8,
                            "paddingTop": 16,
                            "width": 250,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "display": "flex",
                              "height": 40,
                              "justifyContent": "center",
                              "marginBottom": 16,
                            }
                          }
                        >
                          <RNSVGSvgView
                            bbHeight="24"
                            bbWidth="24"
                            focusable={false}
                            height="24"
                            style={
                              Array [
                                Object {
                                  "backgroundColor": "transparent",
                                  "borderWidth": 0,
                                },
                                Object {
                                  "flex": 0,
                                  "height": 24,
                                  "width": 24,
                                },
                              ]
                            }
                            width="24"
                          >
                            <RNSVGGroup>
                              <RNSVGGroup>
                                <RNSVGCircle
                                  cx={12}
                                  cy={12}
                                  fill={4278190080}
                                  propList={
                                    Array [
                                      "fill",
                                    ]
                                  }
                                  r={12}
                                />
                              </RNSVGGroup>
                            </RNSVGGroup>
                          </RNSVGSvgView>
                        </View>
                        <Text
                          numberOfLines={3}
                          style={
                            Object {
                              "fontSize": 15,
                              "height": 54,
                              "lineHeight": 17,
                              "textAlign": "center",
                            }
                          }
                        >
                          Confiscation
                        </Text>
                      </View>
                    </View>
                  </View>
                </CellContainer>
              </AutoLayoutView>
              <CellContainer
                index={-1}
                style={
                  Array [
                    undefined,
                    undefined,
                  ]
                }
              />
              <View
                style={
                  Object {
                    "paddingBottom": 0,
                    "paddingRight": 0,
                  }
                }
              />
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
